homeassistant/components/ohme/button.py

```diff
@@ -0,0 +1,77 @@
+"""Platform for button."""
+
+from __future__ import annotations
+
+from collections.abc import Awaitable, Callable
+from dataclasses import dataclass
+
+from ohme import ApiException, ChargerStatus, OhmeApiClient
+
+from homeassistant.components.button import ButtonEntity, ButtonEntityDescription
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import OhmeConfigEntry
+from .const import DOMAIN
+from .entity import OhmeEntity, OhmeEntityDescription
+
+PARALLEL_UPDATES = 1
+
+
+@dataclass(frozen=True, kw_only=True)
+class OhmeButtonDescription(OhmeEntityDescription, ButtonEntityDescription):
+    """Class describing Ohme button entities."""
+
+    press_fn: Callable[[OhmeApiClient], Awaitable[None]]
+    available_fn: Callable[[OhmeApiClient], bool]
+
+
+BUTTON_DESCRIPTIONS = [
+    OhmeButtonDescription(
+        key="approve",
+        translation_key="approve",
+        press_fn=lambda client: client.async_approve_charge(),
+        is_supported_fn=lambda client: client.is_capable("pluginsRequireApprovalMode"),
+        available_fn=lambda client: client.status is ChargerStatus.PENDING_APPROVAL,
+    ),
+]
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: OhmeConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up buttons."""
+    coordinator = config_entry.runtime_data.charge_session_coordinator
+
+    async_add_entities(
+        OhmeButton(coordinator, description)
+        for description in BUTTON_DESCRIPTIONS
+        if description.is_supported_fn(coordinator.client)
+    )
+
+
+class OhmeButton(OhmeEntity, ButtonEntity):
+    """Generic button for Ohme."""
+
+    entity_description: OhmeButtonDescription
+
+    async def async_press(self) -> None:
+        """Handle the button press."""
+        try:
+            await self.entity_description.press_fn(self.coordinator.client)
+        except ApiException as e:
```

Comment:
What happens if you approve it without a pending request? Does that raise a different error?

Reply:
My thinking was you can't because the button is unavailable then

Reply:
It returns an ApiError at the moment yes, but I'm planning to add a separate exception to the library for it

Reply:
It would be nice to always have the button available and then just raise a proper exception so we can tell the user why this didn't work

Reply:
@joostlek I disagree. I think not letting the user take an action that is currently not possible is better from UX perspective.